plugins/experimental/ssl_session_reuse/src/publish.cc

```diff
@@ -0,0 +1,401 @@
+  }
+
+  TSDebug(PLUGIN, "RedisPublisher::RedisPublisher.PoolRedisConnectTimeout= %d", m_poolRedisConnectTimeout);
+
+  ::pthread_mutex_init(&m_endpointIndexMutex, NULL);
+
+  ::pthread_mutex_init(&m_messageQueueMutex, NULL);
+  ::sem_init(&m_workerSem, 0, 0);
+
+  if (m_redisEndpoints.size() > m_numWorkers) {
+    err = true;
+    TSError("RedisPublisher::RedisPublisher.Number of threads in the thread pool less than the number of redis endpoints");
+  }
+
+  if (!err) {
+    for (int i = 0; i < static_cast<int>(m_numWorkers); i++) {
+      TSThreadCreate(start_worker_thread, static_cast<void *>(this));
+    }
+  }
+}
+
+bool
+RedisPublisher::is_good()
+{
+  return !err;
+}
+
+::redisContext *
+RedisPublisher::setup_connection(const RedisEndpoint &re)
+{
+  ::pthread_t my_id = ::pthread_self();
+  TSDebug(PLUGIN, "RedisPublisher::setup_connection.called by threadId: %d", static_cast<int>(my_id));
+
+  ::redisContext *my_context(NULL);
+  struct ::timeval timeout;
+  timeout.tv_sec  = m_redisConnectTimeout / 1000;
+  timeout.tv_usec = (m_redisConnectTimeout % 1000) * 1000;
+
+  for (int i = 0; i < static_cast<int>(m_redisConnectTries); ++i) {
+    my_context = ::redisConnectWithTimeout(re.m_hostname.c_str(), re.m_port, timeout);
+    if (!my_context) {
+      TSError("RedisPublisher::setup_connection.connect to host: %s and port: %d fail count: %d threadId: %d",
+              re.m_hostname.c_str(), re.m_port, i + 1, static_cast<int>(my_id));
+    } else if (my_context->err) {
+      TSError("RedisPublisher::setup_connection.connect to host: %s port: %d fail count: %d and threadId: %d",
+              re.m_hostname.c_str(), re.m_port, i + 1, static_cast<int>(my_id));
+      teardown_connection(my_context);
+      my_context = NULL;
+    } else {
+      TSDebug(PLUGIN, "RedisPublisher::setup_connection.threadId: %d  successfully connected to the redis instance",
+              static_cast<int>(my_id));
+
+      redisReply *reply = static_cast<redisReply *>(redisCommand(my_context, "AUTH %s", redis_passwd.c_str()));
+
+      if (reply == NULL) {
+        TSError("RedisPublisher::setup_connection. Cannot AUTH redis server, no reply.");
+        teardown_connection(my_context);
+        my_context = NULL;
+      } else if (reply->type == REDIS_REPLY_ERROR) {
+        TSError("RedisPublisher::setup_connection. Cannot AUTH redis server, error reply.");
+        freeReplyObject(reply);
+        teardown_connection(my_context);
+        my_context = NULL;
+      } else {
+        TSDebug(PLUGIN, "RedisPublisher::setup_connection. Successfully AUTH redis server.");
+        freeReplyObject(reply);
+      }
+      break;
+    }
+
+    TSError("RedisPublisher::setup_connection.connect failed.will wait for: %d microseconds and try again.", m_redisRetryDelay);
+    ::usleep(m_redisRetryDelay);
+  }
+
+  return my_context;
+}
+
+void
+RedisPublisher::teardown_connection(::redisContext *context)
+{
+  if (context) {
+    ::redisFree(context);
```

Comment:
Do we need here:

context = nullptr;

Reply:
That does seem like a good idea.

Reply:
Actually this function is gone completely because of the unique_ptr